File: internal/mcpserver/tools_test.go
package mcpserver

import (
	"context"
	"io"
	"net/http"
	"testing"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

// toolCall is one MCP tool exercised over the protocol with realistic
// arguments, and the Domeneshop request it is expected to produce.
//
// Every registered tool has an entry here, and TestEveryToolIsExercised fails
// if a tool is added without one.
type toolCall struct {
	tool string
	args map[string]any

	wantMethod string
	wantPath   string

	status   int
	response string
}

func toolCalls() []toolCall {
	return []toolCall{
		{
			tool: "list_domains", args: map[string]any{},
			wantMethod: "GET", wantPath: "/domains",
			status: 200, response: `[{"id":42,"domain":"example.com"}]`,
		},
		{
			tool: "get_domain", args: map[string]any{"domain_id": 42},
			wantMethod: "GET", wantPath: "/domains/42",
			status: 200, response: `{"id":42,"domain":"example.com"}`,
		},
		{
			tool: "list_dns_records", args: map[string]any{"domain_id": 42},
			wantMethod: "GET", wantPath: "/domains/42/dns",
			status: 200, response: `[{"id":7,"host":"www","type":"A","data":"192.0.2.1"}]`,
		},
		{
			tool: "get_dns_record", args: map[string]any{"domain_id": 42, "record_id": 7},
			wantMethod: "GET", wantPath: "/domains/42/dns/7",
			status: 200, response: `{"id":7,"host":"www","type":"A","data":"192.0.2.1"}`,
		},
		{
			tool:       "create_dns_record",
			args:       map[string]any{"domain_id": 42, "host": "www", "type": "A", "data": "192.0.2.1"},
			wantMethod: "POST", wantPath: "/domains/42/dns",
			status: 201, response: `{"id":7}`,
		},
		{
			tool:       "update_dns_record",
			args:       map[string]any{"domain_id": 42, "record_id": 7, "host": "www", "type": "A", "data": "192.0.2.2"},
			wantMethod: "PUT", wantPath: "/domains/42/dns/7",
			status: 204,
		},
		{
			tool: "delete_dns_record", args: map[string]any{"domain_id": 42, "record_id": 7},
			wantMethod: "DELETE", wantPath: "/domains/42/dns/7",
			status: 204,
		},
		{
			tool: "list_forwards", args: map[string]any{"domain_id": 42},
			wantMethod: "GET", wantPath: "/domains/42/forwards/",
			status: 200, response: `[{"host":"www","frame":false,"url":"https://example.com"}]`,
		},
		{
			tool: "get_forward", args: map[string]any{"domain_id": 42, "host": "www"},
			wantMethod: "GET", wantPath: "/domains/42/forwards/www",
			status: 200, response: `{"host":"www","frame":false,"url":"https://example.com"}`,
		},
		{
			tool:       "create_forward",
			args:       map[string]any{"domain_id": 42, "host": "www", "url": "https://example.com"},
			wantMethod: "POST", wantPath: "/domains/42/forwards/",
			status: 201,
		},
		{
			tool:       "update_forward",
			args:       map[string]any{"domain_id": 42, "host": "www", "url": "https://example.org"},
			wantMethod: "PUT", wantPath: "/domains/42/forwards/www",
			status: 200, response: `{"host":"www","frame":false,"url":"https://example.org"}`,
		},
		{
			tool: "delete_forward", args: map[string]any{"domain_id": 42, "host": "www"},
			wantMethod: "DELETE", wantPath: "/domains/42/forwards/www",
			status: 204,
		},
		{
			tool: "list_invoices", args: map[string]any{},
			wantMethod: "GET", wantPath: "/invoices",
			status: 200, response: `[{"id":1,"status":"paid"}]`,
		},
		{
			tool: "get_invoice", args: map[string]any{"invoice_id": 1},
			wantMethod: "GET", wantPath: "/invoices/1",
			status: 200, response: `{"id":1,"status":"paid"}`,
		},
		{
			tool: "update_dyndns", args: map[string]any{"hostname": "home.example.com", "myip": "192.0.2.1"},
			wantMethod: "GET", wantPath: "/dyndns/update",
			status: 204,
		},
	}
}

// Every tool must pass schema validation with realistic arguments, reach the
// right Domeneshop endpoint, and report a non-error result.
func TestToolCalls(t *testing.T) {
	for _, tc := range toolCalls() {
		t.Run(tc.tool, func(t *testing.T) {
			var gotMethod, gotPath string
			session := connect(t, func(w http.ResponseWriter, r *http.Request) {
				gotMethod, gotPath = r.Method, r.URL.Path
				if tc.status != 0 {
					w.WriteHeader(tc.status)
				}
				_, _ = io.WriteString(w, tc.response)
			})

			res, err := session.CallTool(context.Background(), &mcp.CallToolParams{
				Name: tc.tool, Arguments: tc.args,
			})
			if err != nil {
				t.Fatalf("call failed at the protocol level: %v", err)
			}
			if res.IsError {
				t.Fatalf("tool returned an error: %s", resultText(res))
			}
			if resultText(res) == "" {
				t.Error("tool returned no content")
			}

			if gotMethod != tc.wantMethod {
				t.Errorf("method = %s, want %s", gotMethod, tc.wantMethod)
			}
			if gotPath != tc.wantPath {
				t.Errorf("path = %q, want %q", gotPath, tc.wantPath)
			}
		})
	}
}

// A failing API call must reach the client as a tool error from every tool, so
// no tool can quietly report success on a failed call.
func TestToolCallsReportAPIErrors(t *testing.T) {
	for _, tc := range toolCalls() {
		t.Run(tc.tool, func(t *testing.T) {
			session := connect(t, func(w http.ResponseWriter, r *http.Request) {
				w.WriteHeader(http.StatusForbidden)
				_, _ = io.WriteString(w, `{"code":"forbidden"}`)
			})

			res, err := session.CallTool(context.Background(), &mcp.CallToolParams{
				Name: tc.tool, Arguments: tc.args,
			})
			if err != nil {
				t.Fatalf("call failed at the protocol level: %v", err)
			}
			if !res.IsError {
				t.Fatalf("a 403 was reported as success: %s", resultText(res))
			}
		})
	}
}

// Calling a tool with no arguments at all must fail on the missing required
// ones rather than reaching the API with a zero domain ID.
func TestToolsRequireTheirArguments(t *testing.T) {
	needArgs := map[string]bool{}
	for _, tc := range toolCalls() {
		needArgs[tc.tool] = len(tc.args) > 0
	}

	for tool, required := range needArgs {
		if !required {
			continue
		}
		t.Run(tool, func(t *testing.T) {
			called := false
			session := connect(t, func(w http.ResponseWriter, r *http.Request) {
				called = true
			})

			res, err := session.CallTool(context.Background(), &mcp.CallToolParams{
				Name: tool, Arguments: map[string]any{},
			})
			if err == nil && !res.IsError {
				t.Error("a call with no arguments was accepted")
			}
			if called {
				t.Error("a call with no arguments reached the Domeneshop API")
			}
		})
	}
}

// The table is only meaningful if it covers every registered tool.
func TestEveryToolIsExercised(t *testing.T) {
	session := connect(t, func(w http.ResponseWriter, r *http.Request) {})
	registered := listTools(t, session)

	covered := map[string]bool{}
	for _, tc := range toolCalls() {
		covered[tc.tool] = true
		if _, ok := registered[tc.tool]; !ok {
			t.Errorf("table exercises %q, which is not a registered tool", tc.tool)
		}
	}
	for name := range registered {
		if !covered[name] {
			t.Errorf("tool %q has no entry in the tool call table", name)
		}
	}
}

File: test/mcp/tools.test.ts
import { describe, expect, it } from "vitest";
import { connect, resultText } from "#testing/mcpHarness.js";

describe("domain tools", () => {
  it("lists domains", async () => {
    const harness = await connect((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify([{ id: 1, domain: "example.com" }]));
    });
    try {
      const result = await harness.client.callTool({ name: "list_domains", arguments: {} });

      expect(result.isError).not.toBe(true);
      expect(resultText(result)).toContain("example.com");
      expect(harness.apiServer.requests[0]?.url).toBe("/domains");
    } finally {
      await harness.close();
    }
  });

  it("gets a single domain", async () => {
    const harness = await connect((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ id: 42, domain: "example.com" }));
    });
    try {
      const result = await harness.client.callTool({
        name: "get_domain",
        arguments: { domain_id: 42 },
      });

      expect(result.isError).not.toBe(true);
      expect(resultText(result)).toContain("example.com");
      expect(harness.apiServer.requests[0]?.url).toBe("/domains/42");
    } finally {
      await harness.close();
    }
  });
});

describe("dns record tools", () => {
  it("lists DNS records for a domain", async () => {
    const harness = await connect((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify([{ id: 1, host: "www", type: "A", data: "1.2.3.4" }]));
    });
    try {
      const result = await harness.client.callTool({
        name: "list_dns_records",
        arguments: { domain_id: 42 },
      });

      expect(result.isError).not.toBe(true);
      expect(resultText(result)).toContain("1.2.3.4");
      expect(harness.apiServer.requests[0]?.url).toBe("/domains/42/dns");
    } finally {
      await harness.close();
    }
  });

  it("gets a single DNS record", async () => {
    const harness = await connect((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ id: 7, host: "www", type: "A", data: "1.2.3.4" }));
    });
    try {
      const result = await harness.client.callTool({
        name: "get_dns_record",
        arguments: { domain_id: 42, record_id: 7 },
      });

      expect(result.isError).not.toBe(true);
      expect(resultText(result)).toContain("1.2.3.4");
      expect(harness.apiServer.requests[0]?.url).toBe("/domains/42/dns/7");
    } finally {
      await harness.close();
    }
  });

  it("updates a DNS record, sending every field through to the API", async () => {
    let received: Record<string, unknown> = {};
    const harness = await connect((_req, res, body) => {
      received = JSON.parse(body) as Record<string, unknown>;
      res.writeHead(204);
      res.end();
    });
    try {
      const result = await harness.client.callTool({
        name: "update_dns_record",
        arguments: {
          domain_id: 42,
          record_id: 7,
          host: "www",
          type: "A",
          data: "5.6.7.8",
          ttl: 120,
        },
      });

      expect(result.isError).not.toBe(true);
      expect(harness.apiServer.requests[0]?.method).toBe("PUT");
      expect(harness.apiServer.requests[0]?.url).toBe("/domains/42/dns/7");
      expect(received["data"]).toBe("5.6.7.8");
      expect(received["ttl"]).toBe(120);
    } finally {
      await harness.close();
    }
  });

  it("deletes a DNS record", async () => {
    const harness = await connect((_req, res) => {
      res.writeHead(204);
      res.end();
    });
    try {
      const result = await harness.client.callTool({
        name: "delete_dns_record",
        arguments: { domain_id: 42, record_id: 7 },
      });

      expect(result.isError).not.toBe(true);
      expect(harness.apiServer.requests[0]?.method).toBe("DELETE");
      expect(harness.apiServer.requests[0]?.url).toBe("/domains/42/dns/7");
    } finally {
      await harness.close();
    }
  });
});

describe("forward tools", () => {
  it("lists forwards for a domain", async () => {
    const harness = await connect((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify([{ host: "www", url: "https://example.com", frame: false }]));
    });
    try {
      const result = await harness.client.callTool({
        name: "list_forwards",
        arguments: { domain_id: 42 },
      });

      expect(result.isError).not.toBe(true);
      expect(resultText(result)).toContain("https://example.com");
      expect(harness.apiServer.requests[0]?.url).toBe("/domains/42/forwards/");
    } finally {
      await harness.close();
    }
  });

  it("gets the forward for a specific host", async () => {
    const harness = await connect((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ host: "www", url: "https://example.com", frame: false }));
    });
    try {
      const result = await harness.client.callTool({
        name: "get_forward",
        arguments: { domain_id: 42, host: "www" },
      });

      expect(result.isError).not.toBe(true);
      expect(resultText(result)).toContain("https://example.com");
      expect(harness.apiServer.requests[0]?.url).toBe("/domains/42/forwards/www");
    } finally {
      await harness.close();
    }
  });

  it("updates a forward", async () => {
    let received: Record<string, unknown> = {};
    const harness = await connect((_req, res, body) => {
      received = JSON.parse(body) as Record<string, unknown>;
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ host: "www", url: "https://updated.example.com", frame: true }));
    });
    try {
      const result = await harness.client.callTool({
        name: "update_forward",
        arguments: { domain_id: 42, host: "www", url: "https://updated.example.com", frame: true },
      });

      expect(result.isError).not.toBe(true);
      expect(harness.apiServer.requests[0]?.method).toBe("PUT");
      expect(harness.apiServer.requests[0]?.url).toBe("/domains/42/forwards/www");
      expect(received["frame"]).toBe(true);
      expect(resultText(result)).toContain("updated.example.com");
    } finally {
      await harness.close();
    }
  });

  it("deletes a forward", async () => {
    const harness = await connect((_req, res) => {
      res.writeHead(204);
      res.end();
    });
    try {
      const result = await harness.client.callTool({
        name: "delete_forward",
        arguments: { domain_id: 42, host: "www" },
      });

      expect(result.isError).not.toBe(true);
      expect(harness.apiServer.requests[0]?.method).toBe("DELETE");
      expect(harness.apiServer.requests[0]?.url).toBe("/domains/42/forwards/www");
    } finally {
      await harness.close();
    }
  });
});

describe("invoice tools", () => {
  it("gets a single invoice", async () => {
    const harness = await connect((_req, res) => {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ id: 99, status: "paid", amount: 100 }));
    });
    try {
      const result = await harness.client.callTool({
        name: "get_invoice",
        arguments: { invoice_id: 99 },
      });

      expect(result.isError).not.toBe(true);
      expect(resultText(result)).toContain("paid");
      expect(harness.apiServer.requests[0]?.url).toBe("/invoices/99");
    } finally {
      await harness.close();
    }
  });
});

describe("dynamic DNS tools", () => {
  it("updates a dynamic DNS hostname with an explicit IP", async () => {
    const harness = await connect((_req, res) => {
      res.writeHead(200, { "Content-Type": "text/plain" });
      res.end("good");
    });
    try {
      const result = await harness.client.callTool({
        name: "update_dyndns",
        arguments: { hostname: "home.example.com", myip: "1.2.3.4" },
      });

      expect(result.isError).not.toBe(true);
      expect(resultText(result)).toContain("updated");
      const request = harness.apiServer.requests[0];
      expect(request?.url).toBe("/dyndns/update?hostname=home.example.com&myip=1.2.3.4");
    } finally {
      await harness.close();
    }
  });

  it("omits myip when not provided, letting Domeneshop infer the caller's IP", async () => {
    const harness = await connect((_req, res) => {
      res.writeHead(200, { "Content-Type": "text/plain" });
      res.end("good");
    });
    try {
      const result = await harness.client.callTool({
        name: "update_dyndns",
        arguments: { hostname: "home.example.com" },
      });

      expect(result.isError).not.toBe(true);
      expect(harness.apiServer.requests[0]?.url).toBe("/dyndns/update?hostname=home.example.com");
    } finally {
      await harness.close();
    }
  });
});
